### PR TITLE
Progress #1142 -- Consumables now have their stack count reduced when used

### PR DESCRIPTION
**Pull request by cabeca1143**
* Introduced ItemClass parameter to ItemData in order to differenciate consumables from other items.
* Introduced two GetItem(string itemSpellName) in Inventory and InventoryManager in order to properly assign an item with just it's spellName string
* Introduced a check in HandleCastSpell.cs to check if the spell being casted originates from an item, and another check to see if it is a consumable in order to proceed with the stack removal.

Props to lizardy for his guidance.

Progress #1142

### CODE REVIEW

File: GameServerCore/Domain/IInventoryManager.cs [hunk 1]
```diff
@@ -1,14 +1,17 @@
 ﻿using System.Collections;
+using GameServerCore.Domain.GameObjects;
 using System.Collections.Generic;
 
 namespace GameServerCore.Domain
 {
     public interface IInventoryManager
     {
         IItem GetItem(byte slot);
+        IItem GetItem(string ItemName);
```
Comment (lzardy) [inline GameServerCore/Domain/IInventoryManager.cs]: `ItemName` -> `ItemSpellName`

File: GameServerLib/Items/Inventory.cs [hunk 2]
```diff
@@ -130,7 +143,13 @@ private IItem AddStackingItem(IItemData item)
             }
             return AddNewItem(item);
         }
+        public void RemoveStackingItem(Game game, string itemSpellName, IObjAiBase owner)
+        {
+            IItem item = GetItem(itemSpellName);
 
+            game.PacketNotifier.NotifyRemoveItem(owner, GetItemSlot(item), (byte)(item.StackCount - 1));
```
Comment (lzardy) [inline GameServerLib/Items/Inventory.cs]: Why not Decrement before the packet?

File: GameServerLib/Items/InventoryManager.cs [hunk 3]
```diff
@@ -29,16 +32,31 @@ public IItem GetItem(byte slot)
             return _inventory.GetItem(slot);
         }
 
+        public IItem GetItem(string itemSpellName)
```
Comment (lzardy) [inline GameServerLib/Items/InventoryManager.cs:50]: Replace body of function with `return _inventory.GetItem(itemSpellName);`

File: GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs [hunk 4]
```diff
@@ -35,6 +35,15 @@ public override bool HandlePacket(int userId, CastSpellRequest req)
                 return false;
             }
 
+            if (s.CastInfo.SpellSlot == (int)SpellSlotType.InventorySlots)
+            {
+                var item = s.CastInfo.Owner.Inventory.GetItem(s.SpellName);
+                if (item.ItemData.ItemClass == 1)
```
Comment (lzardy) [inline GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs]: `ItemClass` variable does not reflect consumability.
Use `Consumed` variable instead.

**Review by lzardy — CHANGES_REQUESTED** [review]
Changes needed.

File: GameServerLib/Items/Inventory.cs (hunk 2 above)
Comment (cabeca1143) [reply GameServerLib/Items/Inventory.cs]: Because it would cause a crash with 1 stack, since the item would be removed, resulting on a null exception

Comment (lzardy) [reply GameServerLib/Items/Inventory.cs]: Resolved then.

**Comment by lzardy** [comment]
Fix Tests as well for AppVeyor.

File: GameServerLib/Items/Inventory.cs (hunk 2 above)
Comment (danil179) [inline GameServerLib/Items/Inventory.cs]: PacketNotifier and Game class shouldn't be in Inventory. ~~The packets should be sent after the inventory has been updated in each event (and there are only few: Buy, sell and use).~~

InventoryManger should handle all packet related stuff.

**Review by danil179 — CHANGES_REQUESTED** [review]
Remove Game class everywhere it was added and PacketNotifier from Inventory.

EDIT:
After rechecking the design, I think that the packets should be handled in InventoryManger for all packets related stuff. AddItem has to call BuyItem (add new spells if necessary and modifiers if exist and so on, just copy from Shop) and remove item should do the same. Please only pass PacketManager to this class to avoid propagating Game class everywhere.

File: GameServerLib/Items/Inventory.cs (hunk 2 above)
Comment (lzardy) [reply GameServerLib/Items/Inventory.cs]: Something you could do is add a _toRemove boolean variable to Item.cs, then inside ObjectManager.cs, check each unit with an inventory for removed items, and call the packets there.

Comment (cabeca1143) [reply GameServerLib/Items/Inventory.cs]: So, i found out that it wasn't crashing because the item was null, is was because it was casting a BaseSpell

File: GameServerCore/Domain/IInventoryManager.cs (hunk 1 above)
Comment (cabeca1143) [reply GameServerCore/Domain/IInventoryManager.cs]: resolved

File: GameServerLib/Items/InventoryManager.cs (hunk 3 above)
Comment (cabeca1143) [reply GameServerLib/Items/InventoryManager.cs:50]: resolved

File: GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs (hunk 4 above)
Comment (cabeca1143) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs]: resolved

File: GameServerLib/Items/InventoryManager.cs (hunk 3 above)
Comment (lzardy) [reply GameServerLib/Items/InventoryManager.cs:50]: Not actually resolved.

Comment (cabeca1143) [reply GameServerLib/Items/InventoryManager.cs:50]: Oops my bad

File: GameServerLib/Items/Inventory.cs (hunk 2 above)
Comment (cabeca1143) [reply GameServerLib/Items/Inventory.cs]: resolved

File: GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs [hunk 5]
```diff
@@ -35,6 +35,15 @@ public override bool HandlePacket(int userId, CastSpellRequest req)
                 return false;
             }
 
+            if (s.CastInfo.SpellSlot >= (int)SpellSlotType.InventorySlots && s.CastInfo.SpellSlot < (int)SpellSlotType.BluePillSlot)
```
Comment (lzardy) [inline GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: Change to `== InventorySlots`

Comment (cabeca1143) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: It is >= so it gets everything in-between InventorySlots and BluePillSlot.
Setting it to == InventorySlots will make it only be abble to detect slot 6.

Comment (lzardy) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: Nevermind then ¯\ _(ツ)_/¯

Comment (cabeca1143) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: I just noticed this function is somehow being called twice for some reason

Comment (lzardy) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: It normally does for item actives, although I dont know the cause.

Comment (cabeca1143) [reply GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:38]: "Normally" xD, well at least now i know it wasn't my oopsie hahaha.
I'll investigate it a bit, but i'm just too exausted to dive into it at this point

**Review by danil179 — CHANGES_REQUESTED** [review]
The shop functions for adding/removing items should replace exactly the functions in InventoryManager. There shouldn't be multiple functions to do related things, but instead there should be a simple mechanism to handle them all.

I don't think that `RemoveStackingItem` is really needed if RemoveItem is written correctly. The behavior is the same as the one in Shop.

Also, InventoryManager should at least hold the owner as IObjAiBase (see #527) in the class variables.

Summary:
Two functions in InventoryManager:
bool AddItem(IItem item) [Returns false if the item cannot be added (e.g. inventory is full). The class should handle the stacking automatically if it is consumable by stacking the first item.]
bool RemoveItem(byte slotId, int stacksToRemove=1) [Returns false if the slot was empty at the start. If consumable, removes only stacksToRemove stacks.]
bool RemoveItem(IItemData itemData) [Find the first item to match the data (using GetItem or so) and remove it. Then, returns false if the slot was empty at the start. Should be implemented as return RemoveItem(GetItem(itemData)) or similar]

These two functions (and the helper one) should do the same functionality as the one removed from Shop currently.

**Comment by cabeca1143** [comment]
Judging by the ammount of errors on GameServerLibTest, i'd say changing AddItem to bool wasnt a good idea, i'll work on something else

File: GameServerCore/Domain/IInventoryManager.cs [hunk 6]
```diff
@@ -1,15 +1,18 @@
...
     {
         IItem GetItem(byte slot);
+        IItem GetItem(string ItemSpellName);
         byte GetItemSlot(IItem item);
-        void RemoveItem(byte slot);
-        void RemoveItem(IItem item);
-        IItem AddItem(IItemData item);
+        bool RemoveItem(byte slot, IObjAiBase owner = null, int stacksToRemove = 1);
+        bool RemoveItem(IItem item, IObjAiBase owner = null, int stacksToRemove = 1);
+        Tuple<IItem, bool> AddItem(IItemData item, IObjAiBase owner = null);
```
Comment (lzardy) [inline GameServerCore/Domain/IInventoryManager.cs]: Use `KeyValuePair<INTEL, bool>`

File: GameServerCore/Domain/IStackable.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@ public interface IStackable
         int MaxStacks { get; }
         int StackCount { get; }
         bool IncrementStackCount();
-        bool DecrementStackCount();
+        bool DecrementStackCount(int stacksToRemove = 0);
```
Comment (lzardy) [inline GameServerCore/Domain/IStackable.cs]: Decrement is the opposite of Increment, which goes up by one. Instead of using Decrement, use SetStacks and revert the changes to Decrement.

File: GameServerLib/GameObjects/Other/Stackable.cs [hunk 8]
```diff
@@ -19,14 +19,27 @@ public virtual bool IncrementStackCount()
             return true;
         }
 
-        public virtual bool DecrementStackCount()
+        public virtual bool DecrementStackCount(int stacksToRemove = 0)
```
Comment (lzardy) [inline GameServerLib/GameObjects/Other/Stackable.cs]: Revert and use SetStacks instead.

File: GameServerLib/Items/Inventory.cs [hunk 9]
```diff
@@ -60,14 +75,44 @@ public IItem GetItem(byte slot)
...
+                }
+            }
+            return null;
         }
 
-        public void RemoveItem(IItem item)
+        public void RemoveItem(byte slot, IObjAiBase owner, int stacksToRemove)
         {
-            RemoveItem(GetItemSlot(item));
+            Items[slot].DecrementStackCount(stacksToRemove);
```
Comment (lzardy) [inline GameServerLib/Items/Inventory.cs]: Use SetStacks instead.

File: GameServerLib/Items/InventoryManager.cs [hunk 10]
```diff
@@ -1,22 +1,40 @@
...
 
-        private InventoryManager()
+        private InventoryManager(IPacketNotifier packetNotifier)
         {
+            _packetNotifier = packetNotifier;
             _inventory = new Inventory(this);
         }
 
-        public IItem AddItem(IItemData item)
+        public Tuple<IItem, bool> AddItem(IItemData itemData, IObjAiBase owner = null)
```
Comment (lzardy) [inline GameServerLib/Items/InventoryManager.cs]: Use `KeyValuePair<INTEL, bool>`

**Review by lzardy — CHANGES_REQUESTED** [review]
Some changes needed.

File: GameServerLib/GameObjects/Other/Stackable.cs (hunk 8 above)
Comment (cabeca1143) [reply GameServerLib/GameObjects/Other/Stackable.cs]: Using SetStacks conflicts with the `if (newStacks < 1 || newStacks > MaxStacks)` when setting the stacks to 0 in order to remove the item, i'd have to delete that, if that's alright

Comment (lzardy) [reply GameServerLib/GameObjects/Other/Stackable.cs]: Simply check in RemoveItem before calling SetStacks.

File: GameServerCore/Domain/IInventoryManager.cs (hunk 6 above)
Comment (cabeca1143) [reply GameServerCore/Domain/IInventoryManager.cs]: Resolved

File: GameServerCore/Domain/IStackable.cs (hunk 7 above)
Comment (cabeca1143) [reply GameServerCore/Domain/IStackable.cs]: Resolved

File: GameServerLib/GameObjects/Other/Stackable.cs (hunk 8 above)
Comment (cabeca1143) [reply GameServerLib/GameObjects/Other/Stackable.cs]: Resolved

File: GameServerLib/Items/Shop.cs [hunk 11]
```diff
@@ -56,55 +54,19 @@ public bool HandleItemBuyRequest(int itemId)
                     return false;
                 }
 
-                foreach (var item in ownedItems)
+                foreach (var itens in ownedItems)
```
Comment (danil179) [inline GameServerLib/Items/Shop.cs]: Should be `items`

**Review by danil179 — COMMENTED** [review]
Generally approved. I'll have to test this and check if it works. I'll open an issue about passing the owner as parameter. I think that a class variable is the correct solution (and also merging InventoryManager with Inventory at some point as it is not clear why these two got separated). Is there any scenario that owner is actually needed to be `null`?

File: GameServerLib/Items/Shop.cs (hunk 11 above)
Comment (cabeca1143) [reply GameServerLib/Items/Shop.cs]: > Generally approved. I'll have to test this and check if it works. I'll open an issue about passing the owner as parameter. I think that a class variable is the correct solution (and also merging InventoryManager with Inventory at some point as it is not clear why these two got separated). Is there any scenario that owner is actually needed to be `null`?

I don't remember any actual case in-game they were `null`. I added it because`TestInventoryManager.cs` doesn't contain have any IObjAiBase to be used as an owner, so i thought it would be best to go the "safe" way and make an exeception for it

**Review by danil179 — APPROVED** [review]
Approved. There are many problems with casting and so, but it is not related to the current fix. There are visual problems after consuming the last stack for some unknown reason and the possibility to consume consumables after they have depleted due to this bug. I'll open an issue about an hotfix for that, but the problem is in the casting system and client<-> server miscommunication probably.
